File: new/BTree.go
package main

import (
	"fmt"
)

// Key represents a k-v pair
type Key struct {
	Key, Val int
}

// Node represents a node in the B-tree
type Node struct {
	Leaf     bool
	Keys     []*Key
	Children []*Node
}

// BTree_ represents the B-tree
type BTree_ struct {
	Root *Node
	T    int // Minimum degree
}

// NewKey creates a new Key struct
func NewKey(key, val int) *Key {
	return &Key{key, val}
}

// NewBTreeNode creates a new B-tree node
func NewBTreeNode(leaf bool) *Node {
	return &Node{Leaf: leaf}
}

// NewBTree creates a new B-tree
func NewBTree(t int) *BTree_ {
	return &BTree_{
		Root: NewBTreeNode(true),
		T:    t,
	}
}

// Insert inserts a key into the B-tree
func (bt *BTree_) Insert(key, val int) {
	target := bt.Root
	if len(target.Keys) == (2*bt.T - 1) {
		dest := NewBTreeNode(false)
		bt.Root = dest
		dest.Children = append(dest.Children, target)
		bt.splitChild(dest, 0)
		bt.insertNonFull(dest, key, val)
	} else {
		bt.insertNonFull(target, key, val)
	}
}

// insertNonFull inserts a key into a non-full B-tree node
func (bt *BTree_) insertNonFull(x *Node, key, val int) {
	idx := len(x.Keys) - 1
	if x.Leaf {
		for idx >= 0 && key < x.Keys[idx].Key {
			idx--
		}
		x.Keys = append(x.Keys, nil)
		copy(x.Keys[idx+2:], x.Keys[idx+1:])
		x.Keys[idx+1] = NewKey(key, val)
	} else {
		for idx >= 0 && key < x.Keys[idx].Key {
			idx--
		}
		idx++
		if len(x.Children[idx].Keys) == (2*bt.T - 1) {
			bt.splitChild(x, idx)
			if key > x.Keys[idx].Key {
				idx++
			}
		}
		bt.insertNonFull(x.Children[idx], key, val)
	}
}

// splitChild splits a full child node of a B-tree node
func (bt *BTree_) splitChild(x *Node, idx int) {
	y := x.Children[idx]
	z := NewBTreeNode(y.Leaf)

	x.Children = append(x.Children, nil)
	copy(x.Children[idx+2:], x.Children[idx+1:])
	x.Children[idx+1] = z

	x.Keys = append(x.Keys, nil)
	copy(x.Keys[idx+1:], x.Keys[idx:])
	x.Keys[idx] = y.Keys[bt.T-1]

	z.Keys = append(z.Keys, y.Keys[bt.T:]...)
	y.Keys = y.Keys[:bt.T-1]

	if !y.Leaf {
		z.Children = append(z.Children, y.Children[bt.T:]...)
		y.Children = y.Children[:bt.T]
	}
}

func (bt *BTree_) LookupKey(node *Node, key int) (bool, int) {
	i := 0
	for i < len(node.Keys) && key > node.Keys[i].Key {
		i++
	}

	if i < len(node.Keys) && key == node.Keys[i].Key {
		// Key found in the current node
		return true, key
	} else if node.Leaf {
		// Key not found in a leaf node
		return false, 0
	} else {
		// Recursively search in the appropriate children
		return bt.LookupKey(node.Children[i], key)
	}
}

func (bt *BTree_) PrintTree(n *Node) {
	getVals := func(keys []*Key) (ans [][]int) {
		for _, key := range keys {
			ans = append(ans, []int{key.Key, key.Val})
		}
		return
	}
	q := []*Node{n}
	for len(q) > 0 {
		size := len(q)
		for i := 0; i < size; i++ {
			x := q[0]
			q = q[1:]
			fmt.Print(getVals(x.Keys), "------")
			for _, child := range x.Children {
				q = append(q, child)
			}
		}
		fmt.Println()
	}
}

func (bt *BTree_) PrintTreeWithoutVal(n *Node) {
	getVals := func(keys []*Key) (ans []int) {
		for _, key := range keys {
			ans = append(ans, key.Key)
		}
		return
	}
	q := []*Node{n}
	cnt := 0
	for len(q) > 0 {
		size := len(q)
		for i := 0; i < size; i++ {
			x := q[0]
			q = q[1:]
			fmt.Print(cnt, ":", getVals(x.Keys), "------")
			cnt++
			for _, child := range x.Children {
				q = append(q, child)
			}
		}
		fmt.Println()
	}
}

func main() {
	bTree := NewBTree(4)

	keys := []int{34, 11, 76, 53, 29, 48, 65, 95, 81, 92, 68, 59, 87, 20, 45, 26, 83, 70, 37, 7, 17, 73, 42, 96, 23, 58, 8, 50, 94, 61, 39, 40, 41, 46}

	for _, key := range keys {
		bTree.Insert(key, key*2)
	}

	fmt.Println("B-tree structure:")
	bTree.PrintTreeWithoutVal(bTree.Root)
	fmt.Println()
	fmt.Println()
	fmt.Println()

	fmt.Print("look up 53: ")
	fmt.Println(bTree.LookupKey(bTree.Root, 53))

	fmt.Print("look up 68: ")
	fmt.Println(bTree.LookupKey(bTree.Root, 68))

	fmt.Print("look up 3999:")
	fmt.Println(bTree.LookupKey(bTree.Root, 3999))
}

File: PFS/src/com/neu/nosql/index/BTree.java
package com.neu.nosql.index;

import java.util.LinkedList;
import java.util.Queue;

/**
 * This class represents a B-tree data structure.
 * A B-tree is a self-balancing tree data structure that maintains sorted data and allows searches,
 * sequential access, insertions, and deletions in logarithmic time.
 *
 * The BTree class provides methods for inserting keys into the tree and searching for keys.
 * The tree is implemented as a collection of B-tree nodes, each containing a list of keys and children.
 * The keys are stored in sorted order within each node, and the children represent the subtrees of the node.
 *
 * The B-tree has a minimum degree t, which determines the minimum number of keys in each node.
 * The tree is balanced such that all leaf nodes are at the same level, and each non-leaf node (except the root)
 * contains at least t-1 keys and at most 2t-1 keys.
 *
 * The insert() method inserts a key into the tree, maintaining the properties of the B-tree.
 * The find() method searches for a key in the tree and returns the associated value.
 * The print() method prints the keys of the tree level by level.
 */
public class BTree {

    private static final int MINIMUM_DEGREE = 3;

    private BTreeNode root;
    private final int t; // Minimum degree

    /**
     * Constructs a new B-tree with the default minimum degree.
     */
    public BTree() {
        this.root = new BTreeNode(true);
        this.t = MINIMUM_DEGREE;
    }


    /**
     * Gets the root node of the B-tree.
     */
    public BTreeNode getRoot() {
        return root;
    }

    /**
     * Inserts a key-value pair into the B-tree.
     *
     * @param key the key to be inserted
     * @param val the value associated with the key
     */
    public void insert(int key, int val) {
        BTreeNode r = root;
        if (r.keys.size() == (2 * t - 1)) {
            BTreeNode s = new BTreeNode(false);
            root = s;
            s.children.add(r);
            splitChild(s, 0);
            insertNonFull(s, key, val);
        } else {
            insertNonFull(r, key, val);
        }
    }

    /**
     * Inserts a key-value pair into a non-full node.
     *
     * @param x   the node to insert the key into
     * @param key the key to be inserted
     * @param val the value associated with the key
     */
    private void insertNonFull(BTreeNode x, int key, int val) {
        int i = x.keys.size() - 1;
        if (x.leaf) {
            while (i >= 0 && key < x.keys.get(i).key) {
                i--;
            }
            x.keys.add(i + 1, new Node(key, val));
        } else {
            while (i >= 0 && key < x.keys.get(i).key) {
                i--;
            }
            i++;
            if (x.children.get(i).keys.size() == (2 * t - 1)) {
                splitChild(x, i);
                if (key > x.keys.get(i).key) {
                    i++;
                }
            }
            insertNonFull(x.children.get(i), key, val);
        }
    }

    /**
     * Splits a child node of a given node.
     *
     * @param x the parent node
     * @param i the index of the child to be split
     */
    private void splitChild(BTreeNode x, int i) {
        BTreeNode y = x.children.get(i);
        BTreeNode z = new BTreeNode(y.leaf);

        x.children.add(i + 1, z);
        x.keys.add(i, y.keys.get(t - 1));

        z.keys.addAll(y.keys.subList(t, 2 * t - 1));
        y.keys.subList(t - 1, 2 * t - 1).clear();

        if (!y.leaf) {
            z.children.addAll(y.children.subList(t, 2 * t));
            y.children.subList(t, 2 * t).clear();
        }
    }

    /**
     * Searches for a key in the B-tree and returns the associated value.
     *
     * @param key the key to search for
     * @return the value associated with the key, or -1 if the key is not found
     */
    public int find(int key) {
        return findKey(root, key);
    }

    /**
     * Recursively searches for a key in a B-tree node and its children.
     *
     * @param node the node to search in
     * @param key  the key to search for
     * @return the value associated with the key, or -1 if the key is not found
     */
    public static int findKey(BTreeNode node, int key) {
        int i = 0;
        while (i < node.keys.size() && key > node.keys.get(i).key) {
            i++;
        }

        if (i < node.keys.size() && key == node.keys.get(i).key) {
            // Key found in the current node
            return node.keys.get(i).val;
        } else if (node.leaf) {
            // Key not found in a leaf node
            return -1;
        } else {
            // Recursively search in the appropriate children
            return findKey(node.children.get(i), key);
        }
    }

    /**
     * Prints the keys of the B-tree level by level.
     *
     * @param node the root node of the B-tree
     */
    public void print(BTreeNode node) {
        if (node == null) {
            return;
        }

        Queue<BTreeNode> queue = new LinkedList<>();
        queue.offer(node);

        while (!queue.isEmpty()) {
            int size = queue.size();
            for (int i = 0; i < size; i++) {
                BTreeNode x = queue.poll();
                if (x != null) {
                    // Print keys
                    for (Node key : x.keys) {
                        System.out.print(key + " ");
                    }
                    System.out.print("---");

                    // Enqueue children
                    if (!x.leaf) {
                        for (BTreeNode child : x.children) {
                            queue.offer(child);
                        }
                    }
                }
            }
            System.out.println();
        }
    }
}